sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/CognitiveComplexityWalker.cs

```diff
@@ -0,0 +1,181 @@
+﻿using System;
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using SonarAnalyzer.Common;
+
+namespace SonarAnalyzer.Helpers
+{
+    public class CognitiveComplexityWalker : CSharpSyntaxWalker
+    {
+        private readonly List<SecondaryLocation> secondaryLocations = new List<SecondaryLocation>();
+        private readonly List<ExpressionSyntax> logicalOperationsToIgnore = new List<ExpressionSyntax>();
+
+        private string currentMethodName;
+        private int nestingLevel = 0;
+        private bool hasDirectRecursiveCall = false;
+
+        public int Complexity { get; private set; } = 0;
+        public IEnumerable<SecondaryLocation> SecondaryLocations => secondaryLocations;

```

Comment:
I know it was on my side but maybe we should have a ReadOnlyList in between to ensure other classes can't mess up with secondary locations.